lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1417 @@
+Term *find(Term *x) {
+  if (!x)
+    return nullptr;
+
+  if (auto *var = dyn_cast<VariableTerm>(x)) {
+    if (var->leader == nullptr)
+      return var;
+
+    auto *leader = find(var->leader);
+    if (leader != var->leader)
+      var->leader = leader;
+    return leader;
+  }
+
+  return x;
+}
+
+LogicalResult unify(Term *lhs, Term *rhs);
+
+LogicalResult unify(VariableTerm *x, Term *y) {
+  assert(!x->leader);
+  x->leader = y;
+  return success();
+}
+
+LogicalResult unify(ValueTerm *xv, Term *y) {
+  if (auto *yv = dyn_cast<VariableTerm>(y)) {
+    yv->leader = xv;
+    return success();
+  }
+  if (auto *yv = dyn_cast<ValueTerm>(y)) {
+    return success(xv == yv);
+  }
+  return failure();
+}
+
+// NOLINTNEXTLINE(misc-no-recursion)
+LogicalResult unify(RowTerm *lhsRow, Term *rhs) {
+  if (auto *rhsVar = dyn_cast<VariableTerm>(rhs)) {
+    rhsVar->leader = lhsRow;
+    return success();
+  }
+  if (auto *rhsRow = dyn_cast<RowTerm>(rhs)) {
+    assert(lhsRow->elements.size() == rhsRow->elements.size());
+    for (auto [x, y] : llvm::zip(lhsRow->elements, rhsRow->elements)) {
+      if (failed(unify(x, y)))
+        return failure();
+    }
+    return success();
+  }
+
+  return failure();
+}
+
+// NOLINTNEXTLINE(misc-no-recursion)
+LogicalResult unify(Term *lhs, Term *rhs) {
+  LLVM_DEBUG(auto &out = llvm::errs(); out << "unify x="; dump(out, lhs);
+             out << " y="; dump(out, rhs); out << "\n";);
+  if (!lhs || !rhs)
+    return success();
+  lhs = find(lhs);
+  rhs = find(rhs);
+  if (lhs == rhs)
+    return success();
+  if (auto *lhsVar = dyn_cast<VariableTerm>(lhs))
+    return unify(lhsVar, rhs);
+  if (auto *lhsVal = dyn_cast<ValueTerm>(lhs))
+    return unify(lhsVal, rhs);
+  if (auto *lhsRow = dyn_cast<RowTerm>(lhs))
+    return unify(lhsRow, rhs);
+  return failure();
+}
+
+void solve(Term *lhs, Term *rhs) {
+  auto result = unify(lhs, rhs);
+  (void)result;
+  assert(result.succeeded());
+}
+
+} // namespace
+
+//====--------------------------------------------------------------------------
+// CheckModuleDomains
+//====--------------------------------------------------------------------------
+
+/// Check that a module has complete domain information.
+static LogicalResult checkModuleDomains(GlobalState &globals,
+                                        FModuleLike module) {
+  auto numDomains = globals.circuitInfo.getNumDomains();
+  auto domainInfo = module.getDomainInfoAttr();
+  DenseMap<unsigned, unsigned> typeIDTable;
+  for (size_t i = 0, e = module.getNumPorts(); i < e; ++i) {
+    auto type = module.getPortType(i);
+
+    if (isa<DomainType>(type)) {
+      auto typeID = globals.circuitInfo.getDomainTypeID(domainInfo, i);
+      typeIDTable[i] = typeID;
+      continue;
+    }
+
+    if (auto baseType = type_dyn_cast<FIRRTLBaseType>(type)) {
+      SmallVector<IntegerAttr> associations(numDomains);
+      auto domains = getPortDomainAssociation(domainInfo, i);
+      for (auto index : domains) {
```

Comment:
Nit: inline single-use variables that don't improve readability:


```suggestion
      for (auto index : getPortDomainAssociation(domainInfo, i)) {
```